billing/actions/accounts.py

```diff
@@ -220,3 +229,103 @@ def assign_funds_to_invoice(invoice_id: str) -> bool:
+                                             delinquent
+    :param currency_amount_threshold_map: Balance threshold to consider an user as a
+                                          delinquent.
+                                          Ex: {'CHF': 200, 'EUR': 100, 'NOK': 150}
+    :return: (
+        New delinquent accounts and reason map. Ex: {111: ['Reason why', ...], ...},
+        New compliant accounts ids list
+    )
+    """
+    accounts = Account.objects.filter(id__in=account_ids)
+    new_delinquent_accounts_map = {}
+    compliant_accounts_ids = []
+    for account in accounts:
+        reasons = compute_account_violations(
+            account,
+            unpaid_invoices_threshold,
+            days_since_last_unpaid_threshold,
+            currency_amount_threshold_map,
+        )
+
+        if not account.delinquent and reasons:
+            new_delinquent_accounts_map[account.id] = reasons
+
+        elif account.delinquent and not reasons:
+            compliant_accounts_ids.append(account.id)
+
+    return new_delinquent_accounts_map, compliant_accounts_ids
+
+
+def compute_account_violations(
+    account: Account,
+    unpaid_invoices_threshold: Optional[int],
+    days_since_last_unpaid_threshold: Optional[int],
+    currency_amount_threshold_map: Optional[dict],
+) -> List[str]:
+    """
+    Check if an account has to be marked as delinquent
+    :return: Reasons why account has to be marked as delinquent
+    """
+    reasons = []
+    pending_invoices = account.invoices.filter(status=Invoice.PENDING)
+    if (
+        unpaid_invoices_threshold is not None
+        and pending_invoices.count() >= unpaid_invoices_threshold
+    ):
+        reasons.append(
+            f'Account has more than {unpaid_invoices_threshold} pending invoices'
+        )
+
+    if days_since_last_unpaid_threshold is not None and pending_invoices:
+        last_pending_invoice = pending_invoices.last()
+        days_since_last_pending_invoice = (
+            (date.today() - last_pending_invoice.due_date).days
+        )
+        if days_since_last_pending_invoice >= days_since_last_unpaid_threshold:
+            reasons.append(
+                f'Account has a debt since more than '
+                f'{days_since_last_unpaid_threshold} days'
+            )
+
+    if currency_amount_threshold_map:
+        account_balance = account.balance()
+        for amount_due in account_balance.monies():
+            currency = str(amount_due.currency)
+            if (
+                amount_due.amount < 0
+                and currency in currency_amount_threshold_map
+                and abs(amount_due.amount) > currency_amount_threshold_map[currency]
+            ):
+                reasons.append(
+                    f'Account has a debt of more than {abs(amount_due.amount)} '
+                    f'{currency}'
+                )
+
+    return reasons
+
+
+def swap_delinquent_status(account_ids: List[int]):
```

Comment:
'toggle' instead of 'swap'  (swap is usually between two things, like 1<-->2)

Reply:
Done!